server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java

```diff
@@ -99,40 +115,111 @@ public void copyMessageToMailbox(String message, String sourceMailbox, String de
+        MessageId messageId = getMessagesMethodStepdefs.getMessageId(message);
+        MailboxId destinationMailboxId = mainStepdefs.jmapServer
+            .getProbe(MailboxProbeImpl.class)
+            .getMailbox(MailboxConstants.USER_NAMESPACE, destinationUser, destinationMailbox)
+            .getMailboxId();
+
+        post("[" +
+            "  [" +
+            "    \"setMessages\","+
+            "    {" +
+            "      \"update\": { \"" + messageId.serialize() + "\" : {" +
+            "        \"mailboxIds\": [\"" + destinationMailboxId.serialize() + "\"]" +
+            "      }}" +
+            "    }," +
+            "    \"#0\"" +
+            "  ]" +
+            "]");
+        mainStepdefs.awaitMethod.run();
+    }
+
+    @When("^\"([^\"]*)\" set flags on \"([^\"]*)\" to \"([^\"]*)\"")
+    public void setFlags(String username, String message, List<String> keywords) throws Throwable {
+        userStepdefs.execWithUser(username, () -> setFlags(message, keywords));
+    }
+
     @When("^the user set flags on \"([^\"]*)\" to \"([^\"]*)\"")
     public void setFlags(String message, List<String> keywords) throws Throwable {
-        String username = userStepdefs.getConnectedUser();
         MessageId messageId = getMessagesMethodStepdefs.getMessageId(message);
         String keywordString = keywords
             .stream()
             .map(value -> "\"" + value + "\" : true")
             .collect(Collectors.joining(","));
 
-        Request.Post(mainStepdefs.baseUri().setPath("/jmap").build())
-            .addHeader("Authorization", userStepdefs.getTokenForUser(username).serialize())
-            .bodyString("[" +
-                "  [" +
-                "    \"setMessages\","+
-                "    {" +
-                "      \"update\": { \"" + messageId.serialize() + "\" : {" +
-                "        \"keywords\": {" + keywordString + "}" +
-                "      }}" +
-                "    }," +
-                "    \"#0\"" +
-                "  ]" +
-                "]",
-                ContentType.APPLICATION_JSON)
-            .execute()
-            .discardContent();
+        post("[" +
+            "  [" +
+            "    \"setMessages\","+
+            "    {" +
+            "      \"update\": { \"" + messageId.serialize() + "\" : {" +
+            "        \"keywords\": {" + keywordString + "}" +
+            "      }}" +
+            "    }," +
+            "    \"#0\"" +
+            "  ]" +
+            "]");
         mainStepdefs.awaitMethod.run();
     }
+
+    @Then("^the message \"([^\"]*)\" is not updated$")
+    public void assertIdOfTheFirstMessage(String messageName) throws Exception {
+        MessageId id = getMessagesMethodStepdefs.getMessageId(messageName);
+        assertThat(jsonPath.<Map<String, String>>read("[0][1].notUpdated"))
+            .containsOnlyKeys(id.serialize());
+    }
+
+    private void post(String requestBody) throws Exception {
```

Comment:
why is it private, every tests doing jmap could benefit from that, right ?

Reply:
There is a lot of duplicated code about this accross all JMAP related stapDefs. We would benefit from factorizing this.